internal/typeinfo/typeinfo.go

```diff
@@ -0,0 +1,115 @@
+package typeinfo
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+	"sync"
+)
+
+var cacheMutex sync.RWMutex
+var cache = make(map[reflect.Type]*Info)
+
+// Reflect will return the Info of a given type,
+// generating and caching as required.
+func TypeInfo(value any) (*Info, error) {
+	if value == (any)(nil) {
+		return &Info{}, fmt.Errorf("cannot reflect nil value")
+	}
+
+	v := reflect.ValueOf(value)
+	v = reflect.Indirect(v)
+
+	cacheMutex.RLock()
+	info, found := cache[v.Type()]
+	cacheMutex.RUnlock()
+	if found {
+		return info, nil
+	}
+
+	info, err := generate(v)
+	if err != nil {
+		return &Info{}, err
+	}
+
+	cacheMutex.Lock()
+	cache[v.Type()] = info
+	cacheMutex.Unlock()
+
+	return info, nil
+}
+
+// generate produces and returns reflection information for the input
+// reflect.Value that is specifically required for SQLAir operation.
+func generate(value reflect.Value) (*Info, error) {
+	// Dereference the value if it is a pointer.
+	value = reflect.Indirect(value)
+
+	// Reflection information is only generated for structs.
+	if value.Kind() != reflect.Struct {
+		return &Info{}, fmt.Errorf("cannot reflect type %q, only struct", value.Kind())
+	}
+
+	info := Info{
+		TagToField: make(map[string]Field),
+		FieldToTag: make(map[string]string),
+		Type:       value.Type(),
+	}
+
+	typ := value.Type()
+	for i := 0; i < typ.NumField(); i++ {
+		field := typ.Field(i)
+		// Fields without a "db" tag are outside of SQLAir's remit.
+		tag := field.Tag.Get("db")
+		if tag == "" {
+			continue
+		}
+		tag, omitEmpty, err := parseTag(tag)
+		if err != nil {
+			return &Info{}, fmt.Errorf("cannot parse tag for field %s.%s: %s", typ.Name(), field.Name, err)
+		}
+		info.TagToField[tag] = Field{
+			Name:      field.Name,
+			Index:     i,
+			OmitEmpty: omitEmpty,
+			Type:      reflect.TypeOf(value.Field(i).Interface()),
+		}
+		info.FieldToTag[field.Name] = tag
+	}
+
+	return &info, nil
+}
+
+// This expression should be aligned with the bytes we allow in isNameByte in
+// the parser.
+var validColNameRx = regexp.MustCompile(`^([a-zA-Z_])+([a-zA-Z_0-9])*$`)
+
+// parseTag parses the input tag string and returns its
+// name and whether it contains the "omitempty" option.
+func parseTag(tag string) (string, bool, error) {
+	options := strings.Split(tag, ",")
+
+	var omitEmpty bool
+	// Refuse to parse if there are more than 2 items.
+	if len(options) > 2 {
+		return "", false, fmt.Errorf("too many options in 'db' tag: %s", strings.Join(options, ", "))
+	}
+	if len(options) == 2 {
+		if strings.ToLower(options[1]) != "omitempty" {
+			return "", false, fmt.Errorf("unexpected tag value %q", options[1])
+		}
+		omitEmpty = true
+	}
```

Comment:
Let's please use logic similar to go-yaml here, as we don't really care about how many options we have, but rather about whether the particular flag is supported or not.
```
		inline := false
		fields := strings.Split(tag, ",")
		if len(fields) > 1 {
			for _, flag := range fields[1:] {
				switch flag {
				case "omitempty":
					info.OmitEmpty = true
				case "flow":
					info.Flow = true
				case "inline":
					inline = true
				default:
					return nil, errors.New(fmt.Sprintf("unsupported flag %q in tag %q of type %s", flag, tag, st))
				}
			}
			tag = fields[0]
		}
```

Reply:
Done. Notice that something like this `id_,omitempty,omitempty,omitempty` would be allowed though.

Reply:
That looks awful, but otherwise innocuous. The outcome does exactly what it says on the tin. On the plus side, the suggested logic will tell what flag is wrong, and is extensible to any number of flags.